backend/controller/cronjobs/cronjobs.go

```diff
@@ -0,0 +1,483 @@
+func (s *Service) resetJobsWithNewDeploymentKey(ctx context.Context, deploymentKey optional.Option[model.DeploymentKey]) error {
+	logger := log.FromContext(ctx)
+
+	jobs, err := s.dal.GetCronJobs(ctx)
+	if err != nil {
+		logger.Errorf(err, "failed to get cron jobs")
+		return fmt.Errorf("%s: %w", "failed to get cron jobs", err)
+	}
+	s.jobChanges.Publish(jobChange{
+		changeType:         resetJobs,
+		jobs:               jobs,
+		addedDeploymentKey: deploymentKey,
+	})
+	return nil
+}
+
+func (s *Service) executeJob(ctx context.Context, job dal.CronJob) {
+	logger := log.FromContext(ctx)
+	requestBody := map[string]any{}
+	requestJSON, err := json.Marshal(requestBody)
+	if err != nil {
+		logger.Errorf(err, "could not build body for cron job: %v", job.Key)
+		return
+	}
+
+	req := connect.NewRequest(&ftlv1.CallRequest{
+		Verb: &schemapb.Ref{Module: job.Ref.Module, Name: job.Ref.Name},
+		Body: requestJSON,
+	})
+
+	requestKey := model.NewRequestKey(model.OriginCron, fmt.Sprintf("%s-%s", job.Ref.Module, job.Ref.Name))
+
+	callCtx, cancel := context.WithTimeout(ctx, s.config.Timeout)
+	defer cancel()
+	_, err = s.call(callCtx, req, optional.Some(requestKey), s.originURL.Host)
+	if err != nil {
+		logger.Errorf(err, "failed to execute cron job %v", job.Key)
+	}
+
+	schedule, err := cron.Parse(job.Schedule)
+	if err != nil {
+		logger.Errorf(err, "failed to parse cron schedule %q", job.Schedule)
+		return
+	}
+	next, err := cron.NextAfter(schedule, s.clock.Now().UTC(), false)
+	if err != nil {
+		logger.Errorf(err, "failed to calculate next execution for cron job %v with schedule %q", job.Key, job.Schedule)
+	}
+
+	updatedJob, err := s.dal.EndCronJob(ctx, job, next)
+	if err != nil {
+		logger.Errorf(err, "failed to end cron job %v", job.Key)
+	} else {
+		s.jobChanges.Publish(jobChange{
+			changeType: finishedJobs,
+			jobs:       []dal.CronJob{updatedJob},
+		})
+	}
+}
+
+// killOldJobs looks for jobs that have been executing for too long.
+// A soft timeout should normally occur from the job's context timing out, but there are cases where this does not happen (eg: unresponsive or dead controller)
+// In these cases we need a hard timout after an additional grace period.
+// To do this, this function resets these job's state to idle and updates the next execution time in the db so the job can be picked up again next time.
+func (s *Service) killOldJobs(ctx context.Context) (time.Duration, error) {
+	logger := log.FromContext(ctx)
+	staleJobs, err := s.dal.GetStaleCronJobs(ctx, s.config.Timeout+time.Minute)
+	if err != nil {
+		return 0, err
+	} else if len(staleJobs) == 0 {
+		return time.Minute, nil
+	}
+
+	updatedJobs := []dal.CronJob{}
+	for _, stale := range staleJobs {
+		start := s.clock.Now().UTC()
+		pattern, err := cron.Parse(stale.Schedule)
+		if err != nil {
+			logger.Errorf(err, "Could not kill stale cron job %q because schedule could not be parsed: %q", stale.Key, stale.Schedule)
```

Comment:
Same as above - all of these log messages should be associated with the cron job in the events table so that the user can filter and find issues with their cron jobs.